packages/gatsby/src/schema/__tests__/node-model.js

```diff
@@ -651,43 +715,49 @@ describe(`NodeModel`, () => {
         expect(trackedRootNode).not.toEqual(node)
       })
     })
-
-    describe(`Tracks nodes returned by queries`, () => {
-      it(`Tracks objects when running query without filter`, async () => {
-        const result = await nodeModel.runQuery({
-          query: {},
-          type: schema.getType(`Test`),
-          firstOnly: false,
+    ;[
+      { desc: `with index cache`, cb: () => new Map() }, // Avoids sift
+      { desc: `no index cache`, cb: () => undefined }, // Requires sift
+    ].forEach(({ desc, cb: createIndexCache }) => {
+      describe(`[${desc}] Tracks nodes returned by queries`, () => {
+        it(`Tracks objects when running query without filter`, async () => {
+          const result = await nodeModel.runQuery({
+            query: {},
+            type: schema.getType(`Test`),
+            firstOnly: false,
+            typedKeyValueIndexes: createIndexCache(),
+          })
+
+          expect(result.length).toEqual(2)
+          expect(
+            nodeModel.findRootNodeAncestor(result[0].inlineObject)
+          ).toEqual(result[0])
+          expect(
+            nodeModel.findRootNodeAncestor(result[1].inlineObject)
+          ).toEqual(result[1])
         })
 
-        expect(result.length).toEqual(2)
-        expect(nodeModel.findRootNodeAncestor(result[0].inlineObject)).toEqual(
-          result[0]
-        )
-        expect(nodeModel.findRootNodeAncestor(result[1].inlineObject)).toEqual(
-          result[1]
-        )
-      })
-
-      it(`Tracks objects when running query with filter`, async () => {
-        const result = await nodeModel.runQuery({
-          query: {
-            filter: {
-              inlineObject: {
-                field: {
-                  eq: `fieldOfSecondNode`,
+        it(`Tracks objects when running query with filter`, async () => {
+          const result = await nodeModel.runQuery({
+            query: {
+              filter: {
+                inlineObject: {
+                  field: {
+                    eq: `fieldOfSecondNode`,
+                  },
                 },
               },
             },
-          },
-          type: schema.getType(`Test`),
-          firstOnly: false,
+            type: schema.getType(`Test`),
+            firstOnly: false,
+            typedKeyValueIndexes: createIndexCache(),
```

Comment:
I don't think this does anything, nodeModel gets cache when you create it. (same for all other instances in this code)

Reply:
You're right. I'm not sure why I thought this was necessary. Maybe in another test which also calls runQuery and which was not going through the new code. Will double check all of them.

Reply:
Ah I must have been confused by the tests in `gatsby/src/redux/run-sift.js` (also updated), which call `runSift` directly and which _do_ require the args to be updated.